drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_rtio.c

```diff
@@ -0,0 +1,515 @@
+ * Copyright 2025 CogniPilot Foundation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nxp_lpspi
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(spi_lpspi, CONFIG_SPI_LOG_LEVEL);
+
+#include "spi_nxp_lpspi_priv.h"
+
+struct lpspi_driver_data {
+	struct spi_rtio *rtio_ctx;
+	uint8_t word_size_bytes;
+	struct {
+		size_t words_to_clock;
+		size_t words_clocked_tx;
+		size_t words_clocked_rx;
+	} total;
+	struct {
+		struct rtio_sqe *sqe;
+		size_t words_clocked;
+	} tx_curr;
+	struct {
+		struct rtio_sqe *sqe;
+		size_t words_clocked;
+	} rx_curr;
+	uint8_t lpspi_op_mode;
+};
+
+static inline size_t get_sqe_clock_cycles(struct rtio_sqe *sqe)
+{
+	switch (sqe->op) {
+	case RTIO_OP_RX:
+		return sqe->rx.buf_len;
+	case RTIO_OP_TX:
+		return sqe->tx.buf_len;
+	case RTIO_OP_TINY_TX:
+		return sqe->tiny_tx.buf_len;
+	case RTIO_OP_TXRX:
+		return sqe->txrx.buf_len;
+	default:
+		return 0;
+	}
+}
+
+static inline struct rtio_sqe *get_next_sqe(struct rtio_sqe *sqe)
+{
+	struct rtio_iodev_sqe *curr_iodev_sqe = CONTAINER_OF(sqe, struct rtio_iodev_sqe, sqe);
+	struct rtio_iodev_sqe *next_iodev_sqe = rtio_txn_next(curr_iodev_sqe);
+
+	return &next_iodev_sqe->sqe;
+}
+
+static inline size_t get_total_sqe_clock_cycles(struct rtio_sqe *head)
+{
+	size_t total_size = 0;
+	struct rtio_iodev_sqe *curr_iodev_sqe = CONTAINER_OF(head, struct rtio_iodev_sqe, sqe);
+
+	while (curr_iodev_sqe != NULL) {
+		total_size += get_sqe_clock_cycles(&curr_iodev_sqe->sqe);
+		curr_iodev_sqe = rtio_txn_next(curr_iodev_sqe);
+	}
+
+	return total_size;
+}
+
+static inline const uint8_t *get_sqe_tx_buf(struct rtio_sqe *sqe)
+{
+	switch (sqe->op) {
+	case RTIO_OP_TX:
+		return sqe->tx.buf;
+	case RTIO_OP_TINY_TX:
+		return sqe->tiny_tx.buf;
+	case RTIO_OP_TXRX:
+		return sqe->txrx.tx_buf;
+	default:
+		return NULL;
+	}
+}
+
+static inline uint8_t *get_sqe_rx_buf(struct rtio_sqe *sqe)
+{
+	switch (sqe->op) {
+	case RTIO_OP_RX:
+		return sqe->rx.buf;
+	case RTIO_OP_TXRX:
+		return sqe->txrx.rx_buf;
+	default:
+		return NULL;
+	}
+}
+
+static inline bool lpspi_is_done_tx(const struct device *dev)
+{
+	struct lpspi_data *data = dev->data;
+	struct lpspi_driver_data *lpspi_data = (struct lpspi_driver_data *)data->driver_data;
+
+	return lpspi_data->total.words_clocked_tx >= lpspi_data->total.words_to_clock;
+}
+
+static inline bool lpspi_is_done_rx(const struct device *dev)
+{
+	struct lpspi_data *data = dev->data;
+	struct lpspi_driver_data *lpspi_data = (struct lpspi_driver_data *)data->driver_data;
+
+	return lpspi_data->total.words_clocked_rx >= lpspi_data->total.words_to_clock;
+}
+
+static void lpspi_iodev_complete(const struct device *dev, int status);
+
+static inline uint8_t rx_fifo_cur_len(LPSPI_Type *base)
+{
+	return (base->FSR & LPSPI_FSR_RXCOUNT_MASK) >> LPSPI_FSR_RXCOUNT_SHIFT;
+}
+
+static inline uint8_t tx_fifo_cur_len(LPSPI_Type *base)
+{
+	return (base->FSR & LPSPI_FSR_TXCOUNT_MASK) >> LPSPI_FSR_TXCOUNT_SHIFT;
+}
```

Comment:
should be moved to a common file imo since it is the same between this and the other file

Reply:
Moved to common private header file